api/client.go

```
@@ -45,7 +45,9 @@ func NewClientFromHTTP(httpClient *http.Client) *Client {
 func AddHeader(name, value string) ClientOption {
 	return func(tr http.RoundTripper) http.RoundTripper {
 		return &funcTripper{roundTrip: func(req *http.Request) (*http.Response, error) {
-			req.Header.Add(name, value)
+			if req.Header.Get(name) == "" {
```

why prevent overriding, here? I'd think that repeated calls to `AddHeader` would update a header value, not silently fail to update.

---

It's a good question! The short answer is that I had to add the guard to allow file downloads. Without this change, a file download request would set an `Accept: application/octet-stream`, but additional `Accept` values would get added by our request middleware via AddHeader and AddHeaderFunc, and thus JSON payload would get returned and the download sabotaged.

When I then thought about our AddHeader and AddHeaderFunc utilities, I realized that they are more in the service of adding _default_ request headers (e.g. default Accept, default User-Agent, default Authorization), but if a request already contained any of those headers explicitly, there would be no reason to apply the default.

You have a good point that maybe we should change the verb from "Add" to something else that reflects the current semantics better?

---

ah, that makes sense -- in that case yeah I think we can rename from AddHeader to something like `SetHeaderDefault` or similar

---

https://github.com/cli/cli/pull/1552/files#diff-9e30abd45545c3e865502cae5bf624d9

---

https://github.com/cli/cli/pull/1552#discussion_r491637427

---

> https://github.com/cli/cli/pull/1552#discussion_r491637427

https://github.com/cli/cli/pull/1552/files#diff-9e30abd45545c3e865502cae5bf624d9